storage_controller/src/persistence.rs

```diff
@@ -675,6 +686,130 @@ impl Persistence {
+            move |conn| -> DatabaseResult<_> {
+                Ok(
+                    crate::schema::metadata_health::table
+                        .load::<MetadataHealthPersistence>(conn)?,
+                )
+            },
+        )
+        .await
+    }
+
+    /// Lists all the metadata health records that is unhealthy.
+    #[allow(dead_code)]
+    pub(crate) async fn list_unhealthy_metadata_health_records(
+        &self,
+    ) -> DatabaseResult<Vec<MetadataHealthPersistence>> {
+        use crate::schema::metadata_health::dsl::*;
+        self.with_measured_conn(
+            DatabaseOperation::ListMetadataHealthUnhealthy,
+            move |conn| -> DatabaseResult<_> {
+                Ok(crate::schema::metadata_health::table
+                    .filter(healthy.eq(false))
+                    .load::<MetadataHealthPersistence>(conn)?)
+            },
+        )
+        .await
+    }
+
+    /// Lists all the metadata health records that have not been updated since an `earlier` time.
+    #[allow(dead_code)]
+    pub(crate) async fn list_outdated_metadata_health_records(
+        &self,
+        earlier: chrono::DateTime<chrono::Utc>,
+    ) -> DatabaseResult<
+        Vec<(
+            String,
+            i32,
+            i32,
+            Option<bool>,
+            Option<chrono::DateTime<chrono::Utc>>,
+        )>,
+    > {
+        use crate::schema::metadata_health;
+        use crate::schema::tenant_shards;
+
+        self.with_measured_conn(
+            DatabaseOperation::ListMetadataHealthOutdated,
+            move |conn| -> DatabaseResult<_> {
+                let query = tenant_shards::table
+                    .left_join(
+                        metadata_health::table.on(tenant_shards::tenant_id
+                            .eq(metadata_health::tenant_id)
+                            .and(tenant_shards::shard_number.eq(metadata_health::shard_number))
+                            .and(tenant_shards::shard_count.eq(metadata_health::shard_count))),
+                    )
+                    .select((
+                        tenant_shards::tenant_id,
+                        tenant_shards::shard_number,
+                        tenant_shards::shard_count,
+                        metadata_health::healthy.nullable(),
+                        metadata_health::last_scrubbed_at.nullable(),
+                    ))
+                    .filter(
+                        metadata_health::last_scrubbed_at
+                            .nullable()
+                            // tenant shards that do not have an update
+                            .is_null()
+                            // tenant shards that has outdated metadata health records
+                            .or(metadata_health::last_scrubbed_at.lt(earlier)),
+                    );
+
+                let sql = diesel::debug_query::<diesel::pg::Pg, _>(&query);
+                tracing::info!("Executed query: {}", sql);
```

Comment:
Did you mean to keep this in?

Reply:
56cf5eb02509e8341266a323539e2bbf6256b3ab